finality-aleph/src/sync/substrate/verification/cache.rs

```diff
@@ -0,0 +1,340 @@
+use std::{
+    collections::HashMap,
+    fmt::{Display, Error as FmtError, Formatter},
+};
+
+use aleph_primitives::BlockNumber;
+
+use crate::{
+    session::{first_block_of_session, session_id_from_block_num, SessionId},
+    session_map::AuthorityProvider,
+    sync::substrate::verification::{verifier::SessionVerifier, FinalizationStatus},
+    SessionPeriod,
+};
+
+/// Ways in which a justification can fail verification.
+#[derive(Debug, PartialEq, Eq)]
+pub enum CacheError {
+    UnknownSession,
+    SessionTooOld(SessionId, SessionId),
+    SessionInFuture(SessionId, SessionId),
+}
+
+impl Display for CacheError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use CacheError::*;
+        match self {
+            SessionTooOld(session, lower_bound) => write!(
+                f,
+                "justification from too old session {:?}. Should be at least {:?}",
+                session, lower_bound
+            ),
+            SessionInFuture(session, upper_bound) => write!(
+                f,
+                "justification from session {:?} without known authorities. Should be at most {:?}",
+                session, upper_bound
+            ),
+            UnknownSession => write!(f, "justification from unknown session"),
+        }
+    }
+}
+
+/// Cache storing SessionVerifier structs for multiple sessions. Keeps up to `cache_size` verifiers of top sessions.
+/// If the session is too new or ancient it will fail to return a SessionVerifier.
+/// Highest session verifier this cache returns is for the session after the current finalization session.
+/// Lowest session verifier this cache returns is for `top_returned_session` - `cache_size`.
+pub struct VerifierCache<AP, FS>
+where
+    AP: AuthorityProvider<BlockNumber>,
+    FS: FinalizationStatus,
+{
+    sessions: HashMap<SessionId, SessionVerifier>,
+    session_period: SessionPeriod,
```

Comment:
instead of thi use SessionInfo. It will simplify the code in the impl block.

Reply:
Good idea. I just wonder how exactly this should be done. I would prefer not to have `SessionInfo` as a  parameter of this struct.  Maybe just using implementation parameterized by `BlockNumber` would be enough.

Reply:
why not though?

Reply:
I do not want cache to be dependent on `B: Block` as it is unnecessary. So I would have to either change `SessionInfo` trait to be dependent on `N: AtLeast32BitUnsigned`, or use some struct that has the same methods that is dependent on `BlockNumber`. Also `SessionInfo` always has and will have the same implementation, so I'm not even sure it should be a trait.

Reply:
ok, feel free to resolve this. Although if you would like to do that I am leaving this open ;)

Reply:
I will create a subtask for using `SessionInfo` here and refactoring it.